include/common/result.hpp

```diff
@@ -0,0 +1,146 @@
+template<std::integral T>
+class HasSentinel<T> {
+    static constexpr T value = std::numeric_limits<T>::max();
+};
+
+// Result class.
+// An alternative to std::expected, where an expected value can be contained alongside an unexpected
+// error value.
+// This means that Result will always contain an expected value.
+// Therefore, if the user does not check if a function returned an error, an exception will NOT be
+// thrown, and therefore the thread won't crash.
+template<typename T, CustomError... Errs>
+class Result {
+  public:
+    // Construct a Result with a value and no error value.
+    // Constraint: the value member variable must be able to be constructed with the value
+    // parameter.
+    template<typename U>
+        requires std::constructible_from<T, U>
+    constexpr Result(U&& value)
+        : value(std::forward<U>(value)) {}
+
+    // Construct a Result with a value and an error value.
+    // Constraint: the value member variable must be able to be constructed with the value
+    // parameter.
+    // Constraint: the error parameter must be of a type that may be contained in the error member
+    // variable.
+    template<typename U, CustomError E>
+        requires std::constructible_from<T, U>
+                     && (std::is_same_v<Errs, std::remove_cvref<E>> || ...)
+    constexpr Result(U&& value, E&& error)
+        : value(std::forward<U>(value)),
+          error(std::forward<E>(error)) {}
+
+    // Construct a Result with an error, initializing the normal value to its sentinel value.
+    // Constraint: the type of the normal value must have a specialized sentinel value.
+    // Constraint: the error parameter must be of a type that may be contained in the error member
+    // variable.
+    template<CustomError U>
+        requires Sentinel<T> && (std::is_same_v<Errs, std::remove_cvref<U>> || ...)
+    constexpr Result(U&& error)
+        : value(sentinel_v<T>),
+          error(std::forward<U>(error)) {}
+
+    // Check whether the result contains an error of the given type
+    // Constraint: the custom error type to check for must be able to be contained in the error
+    // member variable.
+    template<CustomError U>
+        requires(std::is_same_v<Errs, std::remove_cvref<U>> || ...)
+    constexpr U has() const& {
+        if (!error.has_value()) {
+            return false;
+        } else {
+            return std::holds_alternative<U>(error);
+        }
+    };
+
+    // return an optional wrapping the given error type.
+    // Constraint: the custom error type to get must be able to be contained in the error
+    // member variable.
+    template<CustomError U>
+        requires(std::is_same_v<Errs, std::remove_cvref<U>> || ...)
+    constexpr std::optional<U> get() const& {
+        if (this->has<U>()) {
+            return std::get<U>(error.value());
+        } else {
+            return std::nullopt;
+        }
+    }
```

Comment:
I probably want to see `get<T>` specialization as well. I know you have the implicit conversions but might as well have that too for symmetry.